grouper/usecases/interfaces.py

```diff
@@ -0,0 +1,47 @@
+"""Interfaces used by use cases to talk to backend services.
+
+Defines general interfaces to use to talk to backend application services (storage, authorization,
+user, group, permission, and so forth) that are shared among multiple use cases.  Also defines the
+exceptions they throw, if needed.
+
+Do not define UI interfaces to talk to frontends here.  There should be a one-to-one correspondance
+between UI interfaces and use cases, so the UI interface is defined in the same file with the use
+case.
+
+By convention, all class names here end in Interface or Exception.
+"""
+
+from abc import ABCMeta, abstractmethod
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from grouper.usecases.authorization import Authorization  # noqa: F401
+
+
+class PermissionNotFoundException(Exception):
+    """Attempt to operate on a permission not found in the storage layer."""
+
+    def __init__(self, name):
+        # type: (str) -> None
+        msg = "Permission {} not found".format(name)
+        super(PermissionNotFoundException, self).__init__(msg)
+
+
+class PermissionInterface(object):
```

Comment:
It looks like we use this in one place. Do we need an interface here? If we do, I'd argue for calling it `PermissionService` and naming the concrete implementations after what differentiates them (e.g. `DatabasePermissionService`)

Reply:
Without the interface, we have a type-checking dependency on the service in the use case, which potentially creates a circular dependency. This is why the hexagonal design pattern always defines an interface in the domain logic for every port between the domain and the application.

Since there's only one instantiation, calling the interface `PermissionService` leads to awkward naming for the actual service (`DefaultPermissionService` or something). I had it that way at one point and thought it was too ugly. I like the clarity of `Interface`, honestly, because it makes clear that this is an abstract base class (which otherwise isn't typical in Python).

Reply:
ah, I see. it feels like this can also be resolved by having the use case interface in a separate package that has no dependencies, but I might be missing something.